### DISCUSSION TITLE
Help with colors variables, theme() function and shadows

### DISCUSSION THREAD

**Discussion by tcolinpa** [tailwindlabs/tailwindcss#11373] [Help] [answered]
Hi, I'm down a rabbit hole and I need a little help.

I've set my primitives colors palette in the theme object, such as 
```
primary: {
  DEFAULT: "#002aff",
  50: "#f2f9ff",
  100: "#c4e5ff",
  200: "#91c6ff",
  300: "#64a1ff",
  400: "#3675ff",
  500: "#002aff",
  600: "#0013ad",
  700: "#000565",
  800: "#000030",
  900: "#01000d",
}
```
then in my base layer I've set the semantics with the theme function, such as
```
--color-brand-minimal: theme(colors.primary.50);
--color-brand-weak: theme(colors.primary.300);
--color-brand-muted: theme(colors.primary.400);
--color-brand-base: theme(colors.primary.DEFAULT);
--color-brand-strong: theme(colors.primary.700);
```
and then I go full circle and link these css variables inside my theme
```
brand: {
  minimal: "var(--color-brand-minimal)",
  weak: "var(--color-brand-weak)",
  muted: "var(--color-brand-muted)",
  base: "var(--color-brand-base)",
  strong: "var(--color-brand-strong)",
},
```

*How can use these colors inside the boxShadow in my theme?*

I tried below
```
boxShadow: {
  "button-neutral-minimal": [
    "0px 1px 2px 0px var(--color-neutral-strong)/[.15]",
    "0px 0px 0px 1px var(--color-neutral-strong)/5",
  ],
  "button-neutral-base": [
    "0px 1px 2px 0px var(--color-neutral-strong)/20",
    "0px 0px 0px 1px var(--color-neutral-strong)/10",
  ],
  "button-neutral-strong": [
    "0px 1px 2px 0px var(--color-neutral-strong)/30",
    "0px 0px 0px 1px var(--color-neutral-strong)/1",
  ],
},
```
but it doesn't work (I already thought it was a long shot)

**Comment by tcolinpa** [comment]
Does the shorthand for opacity property work with hex colors?

**Comment by wongjn** [comment, marked as answer]
Define your primitive colors as RGB (or HSL) primitive numbers, like:

```css
:root {
  --primary-50: 242 249 255;
  --primary-100: 196 229 255;
  /* … */
}
```
Then define your color aliases like:

```css
:root {
  --color-brand-minimal: var(--primary-50);
  --color-brand-weak: var(--primary-300);
  /* … */
}
```

```js
brand: {
  minimal: 'rgb(var(--color-brand-minimal) / <alpha-value>)',
  weak: 'rgb(var(--color-brand-weak) / <alpha-value>)',
},
```
This will allow you to [use opacity modifier in classes](https://tailwindcss.com/docs/customizing-colors#using-css-variables).

Then in your `boxShadow` values, you could either you can use the CSS variables:

```js
boxShadow: {
  "button-neutral-minimal": [
    "0px 1px 2px 0px rgb(var(--color-neutral-strong) / .15)",
    "0px 0px 0px 1px rgb(var(--color-neutral-strong) / 0.5)",
  ],
  // …
},
```

Or use `theme()`:

```
boxShadow: ({ theme }) => ({
  "button-neutral-minimal": [
    `0px 1px 2px 0px ${theme('colors.neutral.strong/15')}`,
    `0px 0px 0px 1px ${theme('colors.neutral.strong/50')}`,
  ],
}),
```
 
See [this Tailwind Play for a proof of concept](https://play.tailwindcss.com/FJvsmgTGUx?file=config).

**Reply by tcolinpa** [reply]
Thanks!

What if I want to use a color from tailwind theme itself that it's a hex?
```
--color-neutral-minimal: theme(colors.gray.50);
--color-neutral-weak: theme(colors.gray.200);
--color-neutral-muted: theme(colors.gray.300);
--color-neutral-base: theme(colors.gray.500);
--color-neutral-strong: theme(colors.gray.900);
```

this theme() function gets the hex code from the native tailwind gray palette, right?
if so, how I can use these? It works just the same?

**Reply by wongjn** [reply]
> What if I want to use a color from tailwind theme itself that it's a hex?

You're out of luck. By abstracting them to CSS variables, Tailwind no longer sees the "raw" values, and only sees `var(--color-neutral-weak)` – it does not know that its value is a hex color.

> this theme() function gets the hex code from the native tailwind gray palette, right?

Right.

> if so, how I can use these?

If you mean "these" by the Tailwind color palette, you could consider not using CSS variables at all:

```js
boxShadow: ({ theme }) => ({
  "button-neutral-minimal": [
    `0px 1px 2px 0px ${theme('colors.gray.900/15')}`,
    `0px 0px 0px 1px ${theme('colors.gray.900/50')}`,
  ],
}),
```

> It works just the same?

No, see above.
